extensions/amp-story/1.0/amp-story-store-service.js

```diff
@@ -389,6 +389,24 @@ export class AmpStoryStoreService {
     }
   }
 
+  /**
+   * Unsubscribes from a state property mutation.
+   * @param {string} key
+   * @param {!Function} listener
+   */
+  unsubscribe(key, listener) {
+    if (!hasOwn(this.state_, key)) {
+      dev().error(TAG, 'Can\'t unsubscribe from unknown state %s.', key);
+      return;
+    }
+    if (!this.listeners_[key]) {
+      dev().error(TAG, 'Can\'t unsubscribe from state with no listeners %s.',
+          listener);
+      return;
+    }
```

Comment:
Not a big deal but I don't think we need to throw an error if there is no listener. The first `if` is interesting because it'd tell you if you were trying to interact with a state that doesn't exist (it actually helped me during development sometimes). 
The store being as flexible as possible, I think we should allow calling `unsubscribe` twice in a row without throwing an error.

Just curious, does this work?

Component A instance 1
`store.subscribe('foo', () => this.bar());`

Component A instance 2
`store.subscribe('foo', () => this.bar());`
`store.unsubscribe('foo', () => this.bar());`

Does it unsubscribe the instance 1 or instance 2?

Reply:
It should unsubscribe from instance 1, since Observable will find and delete the first instance of a given listener.

I added a test to resemble this behavior, and although it is not completely accurate because I'm using two different listeners, it shows the behavior.

Reply:
Ok cool, thanks for adding a test. I found the case I was worried about. Here's one example of how we subscribe to store mutations throughout the codebase:
```
this.storeService_.subscribe(StateProperty.MUTED_STATE, isMuted => {
  this.onMutedStateUpdate_(isMuted);
  this.variableService_.onMutedStateChange(isMuted);
});
```

Which would give this (failing) test:

```
class FakeComponent {
  constructor(callback) {
    this.callback_ = callback;
    storeService.subscribe(StateProperty.BOOKEND_STATE, () => this.callback_());
  }

  unsubscribe() {
    storeService.unsubscribe(StateProperty.BOOKEND_STATE, () => this.callback_());
  }
}

it('should unsubscribe', () => {
  const callbackSpy = sandbox.spy();

  const instance = new FakeComponent(callbackSpy);
  instance.unsubscribe();

  storeService.dispatch(Action.TOGGLE_BOOKEND, true);

  // Will fail, callback gets triggered.
  expect(callbackSpy).to.have.not.been.called;
});
```

To make this work, we'd need to rewrite the way we subscribe to store mutations (first example here) to something like this:

```
onMutedStateUpdate_(isMuted) {
  this.variableService_.onMutedStateChange(isMuted);
  // Rest of the code
}

initializeListeners_() {
  this.storeService_.subscribe(StateProperty.MUTED_STATE, this.onMutedStateUpdate_.bind(this));
}
```

Otherwise it'd silently fail, you'd think you unsubscribed but you didn't... :( I'm worried this pattern is not intuitive given our codebase and might lead to silent bugs. So if we want to keep it you're right, we should throw an error if the listener doesn't exist.

Reply:
On top of this concern, I'm trying to understand the bug you're intending to fix with this change.

You want to call `schedulePause` on page navigation, but you would also need to call this when a user exits the expanded mode (ie: clicks on the X)?
Would a better heuristic be: call `schedulePause` when the component state goes from `EXPANDED` to `HIDDEN`?

This way you wouldn't even need this unsubscribe feature.

Reply:
Yep, the test is failing because you're not passing the same listener on the `subscribe` and on the `unsubscribe`.

If you write it this way the test passes:
```
class FakeComponent {
    constructor(callback) {
      this.callback_ = callback.bind(this);
      storeService.subscribe(StateProperty.BOOKEND_STATE, this.callback_);
    }

    unsubscribe() {
      storeService.unsubscribe(StateProperty.BOOKEND_STATE, this.callback_);
    }
}
```

So yes, we can deal with the concern you mentioned by throwing an error if the listener being unsubscribed from doesn't exist.

Anyone meaning to subscribe/unsubscribe from a StateProperty can follow the pattern mentioned above.

For your second concern, as chatted offline:

I tried calling `schedulePause` when the component state goes from `EXPANDED` to `HIDDEN` but it leads to undesired UX: when the `schedulePause` is ran, it causes the iframe to disappear and reappear in some cases. Kind of like a flicker.

Also, the request from product/UX was to keep playing the video as long as the user stays on the page, and pause it as soon as the user navigates to another page.

Reply:
SG, thanks for explaining : )

So, up to you, we can either:

1. Implement the pattern in this PR. I personally still think it's non intuitive and that we should work around it, but if you feel strongly that it's the best approach let's go with it
2. Figure out in the state update from EXPANDED to HIDDEN if the page changed, and only call schedulePause if it did

Up to you

Reply:
Not sure if I understood option 2, we still want the video to play when we close the embed. And only call `schedulePause` once we change page. That approach would only work when we switch pages while being on the expanded mode.

For (1), did you mean implement that pattern for everyone calling `subscribe` from the store?